fix: categorical legend

hanbyul-here opened the pull request:
**Related Ticket:** #1456 

### Description of Changes
Categorical legend often doesn't have a colormap name but custom colormap values. This PR makes colormap&&rescale combination valid sourceParam. I think I would still like to run it with back-end team re: what makes sourceParams valid. 

### Notes & Questions About Changes
We need to consolidate where we are subbing colormap and rescale value. The current dataflow seems to be scattered, making it difficult to pin down where the source of the data is. IMO, it should happen while reconciling the render data. 

Also it is interesting... that [GeoGLAM dataset](https://earthdata.nasa.gov/dashboard/exploration?datasets=%5B%7B%22id%22%3A%22geoglam%22%2C%22settings%22%3A%7B%22isVisible%22%3Atrue%2C%22opacity%22%3A100%2C%22analysisMetrics%22%3A%5B%7B%22id%22%3A%22mean%22%2C%22label%22%3A%22Average%22%2C%22chartLabel%22%3A%22Average%22%2C%22themeColor%22%3A%22infographicB%22%7D%2C%7B%22id%22%3A%22std%22%2C%22label%22%3A%22St+Deviation%22%2C%22chartLabel%22%3A%22St+Deviation%22%2C%22themeColor%22%3A%22infographicD%22%7D%5D%7D%7D%5D&taxonomy=%7B%7D&search=&date=2025-01-01T05%3A00%3A00.000Z) has been displayed just fine - which is another categorical dataset with colormap value. It just didn't get caught in any of our error Logic - which probably means that we can do better 😬 

### Validation / Testing
Check campfire nlcd data from the preview of https://github.com/NASA-IMPACT/veda-config/pull/669
direct url: https://deploy-preview-669--visex.netlify.app/exploration?datasets=%5B%7B%22id%22%3A%22campfire-nlcd%22%2C%22settings%22%3A%7B%22isVisible%22%3Atrue%2C%22opacity%22%3A100%2C%22analysisMetrics%22%3A%5B%7B%22id%22%3A%22mean%22%2C%22label%22%3A%22Average%22%2C%22chartLabel%22%3A%22Average%22%2C%22themeColor%22%3A%22infographicB%22%7D%2C%7B%22id%22%3A%22std%22%2C%22label%22%3A%22St+Deviation%22%2C%22chartLabel%22%3A%22St+Deviation%22%2C%22themeColor%22%3A%22infographicD%22%7D%5D%7D%7D%5D&taxonomy=%7B%7D&search=&date=2019-01-01T05%3A00%3A00.000Z

anayeaye commented:
I am curious about cases where 
- there are no sourceParams (and none are needed) such as the planet data that is displayed with blue tarp detections https://github.com/NASA-IMPACT/veda-config/blob/develop/datasets/ps_blue_tarp_detections.ej.data.mdx?plain=1#L49
- there are valid sourceParams that have a colormap_name but no rescale (adding rescale would not work) such as soil-texture https://github.com/NASA-IMPACT/veda-config/blob/develop/datasets/soil-texture.data.mdx?plain=1#L28

hanbyul-here commented:
@anayeaye  
Regarding Blue Tarp, this dataset has source param from render extension - https://openveda.cloud/api/stac/collections/blue-tarp-detection  
```
      "rescale": [
        [0, 400]
      ],
      "resampling": "cubic_spline",
      "colormap_name": "reds"
  ```

But if the mdx configuration and render extension are both empty, the dashboard will make requests without any sourceParams.

Re: soil texture data - I think this is the same with geoglam data - it should have thrown an error since it doesn't have a valid sourceparams (at least for the dashboard) - but we have a bug that we are not throwing an error with some specific combination- user configured source param is not valid + render param exists + render param is not valid. This is a bug, but factoring all these cases, there doesn't seem to be any definite rule that we can tell whether the source is valid or not. We can skip any validation process if that reflects the actual rendering of the dataset more correctly?

hanbyul-here commented:
Not directly related but kind of related - do all the datasetes have `dashboard` namespace now? (Can we expect `dashboard` namespace for all the datasets on VEDA?) Asking since we had to depend on user-configured data to retrieve render data ex. https://earth.gov/ghgcenter/api/stac/collections/micasa-carbonflux-monthgrid-v1 If I remember correctly, this dataset render extension did not have `dashboard` namespace before but now it does!

hanbyul-here commented:
Based on the discussion with @anayeaye @smohiudd and @dzole0311  I made several more changes 
- Removed validation
- Changed the precedence of resolving sourceParams. The order is like this: 
// 1. if the layer requires specific asset,
// 1-1. If so, check if renderExtension has that asset as a name space.
// 1-1-1. If so, use the renderExtension data.
// 1-1-2. If not, fallback to user-defined source parameters
// 2. if renderExtension data exists
// 2-1. If so, use the render Extension
// 3. return user defined source parameters (which can be undefined)

@dzole0311 I will appreciate if you can review this one more time 🙇

anayeaye commented:
> // 1-1. If so, check if renderExtension has that asset as a name space.

I think it could be problematic to mix item asset keys with the renders namespace visualization object keys because renders could include multiple assets in source parameters. Can we make a 1:1 mapping of `<dataset-name>.data.mdx layer.id` ([example](https://github.com/NASA-IMPACT/veda-config/blob/develop/datasets/so2.data.mdx?plain=1#L24)) to the renders object keys?

In this case the logic might be 
1. if layer.id in renders keys, use renders[layer.id], 
2. else if renders["dashboard"] exists, use that
3. else use mdx sourceParams

@hanbyul-here @smohiudd @dzole0311

hanbyul-here commented:
thanks for the feedback @anayeaye  - I think eventually we should make an attribute something like `render_namespace` for mdx files, but I am going with layer id for now. Considering some datasets don't have namespaces needed, I made the steps like below.

// 1. Check if render extension has a layer specific namespace from render extension
// 1-1. If so, return the layer specific render data
// 2. Check if user-defined source parameters haave assets defined
// 2-1. If so, return user defined source parameters
// 3. Check if render extension has dashboard namespace
// 3-1. If so, use the render Extension data
// 4. return user defined source parameters (which can be an empty object)

dzole0311 commented:
Tested with a few datasets and works as outlined.

Just to confirm: is there a plan to audit which datasets still rely on [step 2 here](https://github.com/NASA-IMPACT/veda-ui/pull/1462/files#diff-3ef5907984dd01769eba7ea9935db817f59bce65779772a63dfac40ede3a8511R233) and track the progress of moving all of them to backend renders? It sounds like the step can only be safely removed once we have confidence that no datasets rely on mdx settings anymore. 

Is there maybe a way to log when step 2 is used in the UI (e.g, logging the dataset / layer to an endpoint), so we have visibility into which datasets/layers are missing render extensions?

anayeaye commented:
@hanbyul-here @dzole0311 this change looks like it would support all existing mdx configurations and would incrementally start using the renders info from STAC 😎. Right now there are no custom namespaced renders configs but the majority of the production collections do have a dashboard namespace. 

> 1. Check if render extension has a layer specific namespace from render extension

We do not have a task in our backlog to update VEDA collections with namespaced renders configuration but we could encourage namespaced renders going forward and then plan to update the existing collections (might need to make this an objective for the next PI if it is the plan).


> 2. Check if user-defined source parameters haave assets defined

I'm a little unclear what has assets/ has no assets is used for. Is it because contributors were not previously required to set an asset in the mdx file because veda-ui assumed there would be an asset with the key `cog_default`? I think [that is what is happening](https://github.com/search?q=repo%3ANASA-IMPACT%2Fveda-ui%20cog_default&type=code) but I don't know what this check is doing.

cc: @smohiudd

hanbyul-here commented:
@anayeaye  - yah, it does fallback to cog_default if asset is not defined. But if an asset is defined, it usually means that the asset requires a different visualization than default (dashboard name spaced one)  - such as soil texture data example (https://openveda.cloud/api/stac/collections/soil-texture). So this is a work around, not perfect, but can't think of any other way now. I think we should have a PI objective, something like making the back-end the source of truth for dashboard rendering and work together!

slesaad commented on app/scripts/components/exploration/data-utils-no-faux-module.ts:
```diff
@@ -201,48 +199,45 @@ function flattenAndCalculateMinMax(rescale: number[]): [number, number] {
+export type SourceParametersWithLayerId = SourceParameters & {
+  layerId: string;
+};
+// renderParams precedence
+// 1. Check if render extension has a layer specific namespace from render extension
+// 1-1. If so, return the layer specific render data
+// 2. Check if user-defined source parameters haave assets defined
+// 2-1. If so, return user defined source parameters
+// 3. Check if render extension has dashboard namespace
+// 3-1. If so, use the render Extension data
+// 4. return user defined source parameters (which can be an empty object)
```
Sorry, I'm just now seeing this @hanbyul-here 

I'd recommend that the user-defined `sourceParams` be # 1 in precedence.

The renders that the stac catalog provides are the data provider's recommendation. However, the user who is working on adding the dataset to the dashboard should be able to customize how the data is visualized rather than forcing them to use whatever is recommended.

hanbyul-here replied:
What you suggested makes a lot of sense. This approach mainly comes from our short-term goal of delegating as much as possible to the STAC endpoint while keeping MDX metadata as lightweight as possible.

From my understanding, at least for the dashboard instance, data providers also edit the data. This means they are duplicating efforts—first ingesting the data and then writing MDX metadata. Is there any scenario where authors are meaningfully overwriting sourceParameters? For example, in the recent bug with multiple bidx, authors were overwriting the source parameters with asset_bidx, but ultimately, this was just a rewrite of bidx values.

The best approach might be to allow users to overwrite only specific attributes rather than the entire source parameter object - so they don't have to scratch their heads when they just want to see the data in a different colormap.
